src/superstring/core.clj

```diff
@@ -29,52 +32,52 @@
   "Create an alias for all public vars in ns in this ns."
   [namespace]
   `(do ~@(map
-          (fn [n] `(defalias ~(.sym n) ~(symbol (str (.ns n)) (str (.sym n)))))
+          (fn [^clojure.lang.Var n] `(defalias ~(.sym n) ~(symbol (str (.ns n)) (str (.sym n)))))
           (vals (ns-publics namespace)))))
 
 (alias-ns clojure.string)
-(def substring clojure.core/subs)
+(def ^{:tag String} substring clojure.core/subs)
 
-(defn ^Long length
+(defn length
   "Return the length of s."
-  [^String s]
-  {:pre [(string? s)]
-   :post [(integer? %)]}
-  (.length s))
+  (^long [^String s]
+    {:pre [(string? s)]
+     :post [(integer? %)]}
+    (.length s)))
 
-(defn ^Long index-of
+(defn index-of
   "Return the starting position of the first occurrence of needle in s or nil.
 
   If start is provided, start the search at that position in s."
-  ([^String s needle]
+  (^Long [^String s needle]
    {:pre [(string? s) (string? needle)]
     :post [(or (integer? %) (nil? %))]}
    (index-of s needle 0))
-  ([^String s ^String needle start]
+  (^Long [^String s ^String needle ^long start]
    {:pre [(string? s) (string? needle) (integer? start)]
     :post [(or (integer? %) (nil? %))]}
-   (let [i (.indexOf s needle start)]
+   (let [i (long (.indexOf s needle (int start)))]
```

Comment:
The JVM optimises this extremely well (I think it is virtually free on modern CPUs) so it doesn't add any noticeable overhead. 

I think it was also necessary in this case to avoid an error when casting to the `Long` return value.